### ISSUE TITLE
First non-default bridge network created has non-functional nameserver in multi-network case

### ISSUE THREAD

**Issue by corrieb** [vmware/vic#5660]
**User Statement:**

As a container developer, I need containers to be able to refer to each other using their name if they're deployed to the same custom network. This is documented here: https://vmware.github.io/vic-product/assets/files/html/1.1/vic_app_dev/network_use_cases.html#newbridge

Inconsistent name resolution will cause application failures on deployment.

**Details:**

If I run the following series of commands with a fresh VCH and try to ping the name of the container, it always fails.

```
docker network create bridge1
docker network create bridge2
docker create -it --name test1 --net bridge1 bensdoings/ifconfig
docker network connect bridge2 test1
docker start -ai test1
     $ ping test1
        ping: unknown host test1
```
However, if I install a fresh VCH and create and delete a bridge network before running the above commands, it works:

```
docker network create foo
docker network rm foo
docker network create bridge1
docker network create bridge2
docker create -it --name test1 --net bridge1 bensdoings/ifconfig
docker network connect bridge2 test1
docker start -ai test1
     $ ping test1
        PING test1 (172.18.0.2) 56(84) bytes of data.
        64 bytes from bridge1.localhost (172.18.0.2): icmp_seq=1 ttl=64 time=0.012 ms
        64 bytes from bridge1.localhost (172.18.0.2): icmp_seq=2 ttl=64 time=0.023 ms
```
There is something about the initialization of that first network that causes problems. I've tried this multiple times with the same results.

**Acceptance Criteria:**

This is a significant problem. It will severely impact the ability to deploy multi-container applications. I would suggest this should be high priority to fix.

We must have a regression test for this.

**Comment by corrieb** [comment]
[container-logs-5660.zip](https://github.com/vmware/vic/files/1129147/container-logs-5660.zip)

**Comment by corrieb** [comment]
More information on this issue. It seems that the name resolution may or may not work in the same container by just trying ping multiple times. In this case, you can see how the same ping command worked and then failed to work. When it did work, it took quite a few seconds to resolve the name. 

```
$ docker network create bridge1
1777dcc608b7316078abbfb54545901b6532395184625221e1947cedc8e65541
$ docker network create bridge2
727d754f738e9c7935b01541cff8b734c3ef89949c78f71e4300160db5f68015
$ docker create -it --name test1 --net bridge1 bensdoings/ifconfig
410ad26cd7f44c2f4bd8d18a2488474057bb2b5b5d8ea4501f9d41824e904015
$ docker network connect bridge2 test1
$ docker start -ai test1
root@410ad26cd7f4:/# ping test1
ping: unknown host test1
root@410ad26cd7f4:/# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:b4:c9:83 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet 172.18.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fc00:10:118:69:250:56ff:feb4:c983/64 scope global mngtmpaddr dynamic 
       valid_lft 2591975sec preferred_lft 604775sec
    inet6 fc00:10:118:68:250:56ff:feb4:c983/64 scope global mngtmpaddr dynamic 
       valid_lft 2591975sec preferred_lft 604775sec
    inet6 fe80::250:56ff:feb4:c983/64 scope link 
       valid_lft forever preferred_lft forever
root@410ad26cd7f4:/# more /etc/resolv.conf 
nameserver 172.17.0.1
nameserver 172.18.0.1
options timeout:15
options attempts:5
root@410ad26cd7f4:/# ping test1
PING test1 (172.18.0.2) 56(84) bytes of data.
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=1 ttl=64 time=0.021 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=2 ttl=64 time=0.045 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=3 ttl=64 time=0.031 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=4 ttl=64 time=0.024 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=5 ttl=64 time=0.036 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=6 ttl=64 time=0.025 ms
^C
--- test1 ping statistics ---
6 packets transmitted, 6 received, 0% packet loss, time 4999ms
rtt min/avg/max/mdev = 0.021/0.030/0.045/0.009 ms
```
Now I started another container in another shell:

```
$ docker run -it --name test2 --net bridge2 bensdoings/ifconfig

root@3f9acb25b16a:/# 
root@3f9acb25b16a:/# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:b4:96:2c brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.3/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fc00:10:118:69:250:56ff:feb4:962c/64 scope global mngtmpaddr dynamic 
       valid_lft 2591992sec preferred_lft 604792sec
    inet6 fc00:10:118:68:250:56ff:feb4:962c/64 scope global mngtmpaddr dynamic 
       valid_lft 2591992sec preferred_lft 604792sec
    inet6 fe80::250:56ff:feb4:962c/64 scope link 
       valid_lft forever preferred_lft forever
root@3f9acb25b16a:/# ping test1
PING test1 (172.18.0.2) 56(84) bytes of data.
64 bytes from 172.18.0.2: icmp_seq=1 ttl=64 time=0.854 ms
64 bytes from 172.18.0.2: icmp_seq=2 ttl=64 time=0.445 ms
^C
--- test1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.445/0.649/0.854/0.206 ms
root@3f9acb25b16a:/# ping test2
PING test2 (172.18.0.3) 56(84) bytes of data.
64 bytes from bridge2.localhost (172.18.0.3): icmp_seq=1 ttl=64 time=0.013 ms
64 bytes from bridge2.localhost (172.18.0.3): icmp_seq=2 ttl=64 time=0.042 ms
^C
--- test2 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1000ms
rtt min/avg/max/mdev = 0.013/0.027/0.042/0.015 ms
root@3f9acb25b16a:/# more /etc/resolv.conf 
nameserver 172.18.0.1
options timeout:15
options attempts:5
root@3f9acb25b16a:/# 
```

The pings in this container are as responsive as I would expect and I can ping both containers. I then went back to the first.

```
root@410ad26cd7f4:/# ping test2
PING test2 (172.18.0.3) 56(84) bytes of data.
64 bytes from 172.18.0.3: icmp_seq=1 ttl=64 time=0.616 ms
64 bytes from 172.18.0.3: icmp_seq=2 ttl=64 time=0.713 ms
64 bytes from 172.18.0.3: icmp_seq=3 ttl=64 time=0.506 ms
64 bytes from 172.18.0.3: icmp_seq=4 ttl=64 time=0.475 ms
64 bytes from 172.18.0.3: icmp_seq=5 ttl=64 time=0.506 ms
^C
--- test2 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 18036ms
rtt min/avg/max/mdev = 0.475/0.563/0.713/0.090 ms
root@410ad26cd7f4:/# ping test2
ping: unknown host test2
root@410ad26cd7f4:/# ping test2
PING test2 (172.18.0.3) 56(84) bytes of data.
64 bytes from 172.18.0.3: icmp_seq=1 ttl=64 time=0.583 ms
64 bytes from 172.18.0.3: icmp_seq=2 ttl=64 time=0.711 ms
64 bytes from 172.18.0.3: icmp_seq=3 ttl=64 time=0.482 ms
64 bytes from 172.18.0.3: icmp_seq=4 ttl=64 time=0.450 ms
64 bytes from 172.18.0.3: icmp_seq=5 ttl=64 time=0.479 ms
64 bytes from 172.18.0.3: icmp_seq=6 ttl=64 time=0.436 ms
64 bytes from 172.18.0.3: icmp_seq=7 ttl=64 time=0.532 ms
^C
--- test2 ping statistics ---
7 packets transmitted, 7 received, 0% packet loss, time 20016ms
rtt min/avg/max/mdev = 0.436/0.524/0.711/0.093 ms
root@410ad26cd7f4:/# 
```

Note how long it takes - almost a 10 second delay before succeeding and then on one occasion it didn't work at all

**Comment by corrieb** [comment]
Subsequent test, it seems like the name server in bridge1 has just stopped functioning:

```
$ docker create -it --name test1 --net bridge1 bensdoings/ifconfig
4d2decf6dfe24f7773605f1cbac4b5210b337139f979186d8606f6250d6d5915
$ docker start -ai test1
root@4d2decf6dfe2:/# ping test1
ping: unknown host test1
root@4d2decf6dfe2:/# ping test1
ping: unknown host test1
root@4d2decf6dfe2:/# exit
exit
$ docker rm test1
test1
$ docker create -it --name test1 --net bridge2 bensdoings/ifconfig
a16371f45d3b4843a344f422af391c6df08d95dcd9e217422bb9d63772ca0e52
$ docker start -ai test1
root@a16371f45d3b:/# ping test1
PING test1 (172.18.0.2) 56(84) bytes of data.
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=1 ttl=64 time=0.012 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=2 ttl=64 time=0.035 ms
64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=3 ttl=64 time=0.022 ms
q64 bytes from bridge2.localhost (172.18.0.2): icmp_seq=4 ttl=64 time=0.036 ms
^C
--- test1 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 2999ms
rtt min/avg/max/mdev = 0.012/0.026/0.036/0.010 ms
```

**Comment by corrieb** [comment]
Note there are subsequent tests that can be run with multiple containers, but this is the most basic manifestation of the problem

**Comment by hickeng** [comment]
We have [this test](https://github.com/vmware/vic/blob/master/tests/test-cases/Group1-Docker-Commands/1-17-Docker-Network-Connect.robot#L55) that exercises this same path, but uses two containers for the ping.
It's possible this is very niche and specifically impacts resolution of a containers own hostname.

**Comment by corrieb** [comment]
What's weird @hickeng is that it only happens with that first created network. All subsequent networks created seem to work fine. I don't think there's anything necessarily wrong with the configuration here.

**Comment by chengwang86** [comment]
@corrieb Have you tried your two examples with vanilla docker? 

Neither of these examples worked in my local testing.

**Comment by chengwang86** [comment]
Currently I cannot fully reproduce this issue due to https://github.com/vmware/vic/issues/5989. But I can bypass it by manually adding firewall rules to allow loopback ping, so that I can keep working on this ticket.

Fix for this ticket should go in after #5989 is fixed.

**Comment by chengwang86** [comment]
Update:

I tried the following on vic:
```
docker network create bridge1
docker network create bridge2
docker create -it --name test1 --net bridge1 busybox
docker network connect bridge2 test1
docker start -ai test1
```
Then within the container,
```
/ # ping test1
PING test1 (172.18.0.2): 56 data bytes
ping: sendto: Operation not permitted
```
So instead of the `ping: unknown host test1` from the original post, it seems like the nameserver is able to resolve `test1` to `172.18.0.2`. 

Here `ping: sendto: Operation not permitted` is due to #5989

**Comment by chengwang86** [comment]
After I manually added the `loopback` related firewall rules by the following cmds (allowing input/output traffic on loopback):
```
/ # chroot /.tether /iptables -I VIC 1 -i lo -j ACCEPT
/ # chroot /.tether /iptables -I VIC 1 -o lo -j ACCEPT
```
`ping test1` works even in the first example Ben posted. So it seems like this issue can be resolved automatically if #5989 is fixed. 

Now I just need to add integration test for this issue.
